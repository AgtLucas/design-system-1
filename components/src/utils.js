
export function getTextWidth(text, font) {
  const canvas = getTextWidth.canvas || (getTextWidth.canvas = document.createElement("canvas"));
  const context = canvas.getContext("2d");
  context.font = font;
  const metrics = context.measureText(text);
  return metrics.width;
}

export const addPx = (val1, val2) => {
  const val2Str = !val2 ? "1px" : `${val2}`;
  return (`${parseInt(val1.replace("px", ""), 10) + parseInt(val2Str.replace("px", ""), 10)}px`);
};

export const subPx = (val1, val2) => {
  const val2Str = !val2 ? "1px" : `${val2}`;
  return (`${parseInt(val1.replace("px", ""), 10) - parseInt(val2Str.replace("px", ""), 10)}px`);
};

export const multPx = (val1, val2) => {
  const val2Str = !val2 ? "1px" : `${val2}`;
  return (`${Math.round(parseFloat(val1.replace("px", "")) * parseFloat(val2Str.replace("px", "")))}px`);
};
