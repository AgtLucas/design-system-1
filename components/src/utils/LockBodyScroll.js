import React from "react";

class LockBodyScroll extends React.Component {
  componentDidMount() {
    this.initialBodyOverflow = document.body.style.overflow;
    document.body.style.overflow = "hidden";
  }

  componentWillUnmount() {
    document.body.style.overflow = this.initialBodyOverflow;
  }

  render() {
    return null;
  }
}

export default LockBodyScroll;
