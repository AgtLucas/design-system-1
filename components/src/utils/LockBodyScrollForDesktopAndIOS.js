import React from "react";

const preventIOSScroll = e => {
  e.preventDefault();
};

class LockBodyScrollForDesktopAndIOS extends React.Component {
  constructor() {
    super();
    const initialBodyOverflow = undefined;
  }

  componentDidMount() {
    this.initialBodyOverflow = document.body.style.overflow;
    document.body.style.overflow = "hidden";
    document.body.addEventListener("touchmove", preventIOSScroll, { passive: false });
  }

  componentWillUnmount() {
    document.body.style.overflow = this.initialBodyOverflow;
    document.body.removeEventListener("touchmove", preventIOSScroll, { passive: false });
  }

  render() {
    return null;
  }
}

export default LockBodyScrollForDesktopAndIOS;
