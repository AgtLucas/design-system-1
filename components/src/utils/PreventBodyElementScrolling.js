import React from "react";
import PropTypes from "prop-types";

class PreventBodyElementScrolling extends React.Component {
  componentDidMount() {
    this.initialBodyOverflow = document.body.style.overflow;
    document.body.style.overflow = "hidden";
  }

  componentWillUnmount() {
    document.body.style.overflow = this.initialBodyOverflow;
  }

  render() {
    const { children } = this.props;
    return <>{children}</>;
  }
}

PreventBodyElementScrolling.propTypes = {
  children: PropTypes.node
};

PreventBodyElementScrolling.defaultProps = {
  children: null
};
export default PreventBodyElementScrolling;
