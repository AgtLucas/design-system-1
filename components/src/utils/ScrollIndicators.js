import React from "react";
import { Icon } from "../Icon";
import theme from "../theme";

const ScrollIndicatorTop = () => (
  <div
    style={{
      position: "absolute",
      top: 0,
      left: "calc(50% - 32px)",
      right: "calc(50% - 32px)",
      height: "32px",
      width: "64px",
      background: theme.colors.lightGrey,
      opacity: 0.8,
      borderRadius: "0 0 16px 16px",
      pointerEvents: "none"
    }}
  >
    <Icon style={{ display: "block", margin: "0 auto" }} color="darkGrey" icon="upArrow" />
  </div>
);

const ScrollIndicatorBottom = () => (
  <div
    style={{
      position: "absolute",
      bottom: 1,
      left: "calc(50% - 32px)",
      right: "calc(50% - 32px)",
      height: "32px",
      width: "64px",
      background: theme.colors.lightGrey,
      opacity: 0.8,
      borderRadius: "16px 16px 0 0",
      pointerEvents: "none"
    }}
  >
    <Icon style={{ display: "block", margin: "8px auto" }} color="darkGrey" icon="downArrow" />
  </div>
);

class ScrollIndicators extends React.Component {
  constructor() {
    super();
    this.state = {
      scrollTop: 0
    };
    this.menuRef = React.createRef();
    this.handleScroll = this.handleScroll.bind(this);
  }

  handleScroll() {
    if (this.menuRef.current) {
      this.setState({
        scrollTop: this.menuRef.current.scrollTop
      });
    }
  }

  contentHiddenBelow() {
    if (this.menuRef.current) {
      return this.state.scrollTop + this.menuRef.current.offsetHeight < this.menuRef.current.scrollHeight;
    } else {
      return false;
    }
  }

  contentHiddenAbove() {
    if (this.menuRef.current) {
      return this.state.scrollTop !== 0 && this.menuRef.current.offsetHeight < this.menuRef.current.scrollHeight;
    } else {
      return false;
    }
  }

  render() {
    return (
      <div style={{ position: "relative" }}>
        {this.contentHiddenAbove() && <ScrollIndicatorTop />}
        {React.cloneElement(this.props.children, {
          ref: this.menuRef,
          onScroll: this.handleScroll
        })}
        {this.contentHiddenBelow() && <ScrollIndicatorBottom />}
      </div>
    );
  }
}

export default ScrollIndicators;
