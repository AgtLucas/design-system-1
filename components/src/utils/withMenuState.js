import React from "react";
import PropTypes from "prop-types";
import { keyCodes } from "../constants";

class MenuState extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: this.props.defaultOpen
    };

    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.toggleMenu = this.toggleMenu.bind(this);
    this.openMenu = this.openMenu.bind(this);
    this.closeMenu = this.closeMenu.bind(this);
  }

  componentDidMount() {
    document.addEventListener("keydown", this.handleKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener("keydown", this.handleKeyDown);
    this.clearScheduled();
  }

  setMenuState(newState, skipTimer = true) {
    const { showDelay, hideDelay } = this.props;

    this.clearScheduled();
    this.conditionallyApplyDelay(
      () => this.setState({ isOpen: newState }),
      skipTimer,
      newState ? showDelay : hideDelay
    );
  }

  toggleMenu(skipTimer = true) {
    const { isOpen } = this.state;
    const { showDelay, hideDelay } = this.props;

    this.clearScheduled();
    this.conditionallyApplyDelay(
      () => this.setState(prevState => ({ isOpen: !prevState.isOpen })),
      skipTimer,
      isOpen ? hideDelay : showDelay
    );
  }

  clearScheduled() {
    clearTimeout(this.timeoutID);
  }

  conditionallyApplyDelay(fnc, skipTimer, delay) {
    if (!skipTimer) {
      this.timeoutID = setTimeout(fnc, delay);
    } else {
      fnc();
    }
  }

  closeMenu(skipTimer) {
    this.setMenuState(false, skipTimer);
  }

  openMenu(skipTimer) {
    this.setMenuState(true, skipTimer);
  }

  handleKeyDown(event) {
    switch (event.keyCode) {
      case keyCodes.ESC:
        this.closeMenu();
        break;
      default:
        break;
    }
  }

  render() {
    const { isOpen } = this.state;
    const { children: renderMenu } = this.props;

    return renderMenu({
      isOpen,
      toggleMenu: this.toggleMenu,
      handleMenuKeydown: this.handleKeyDown,
      openMenu: this.openMenu,
      closeMenu: this.closeMenu
    });
  }
}

MenuState.propTypes = {
  children: PropTypes.func.isRequired,
  showDelay: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hideDelay: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  defaultOpen: PropTypes.bool
};

MenuState.defaultProps = {
  showDelay: 0,
  hideDelay: 0,
  defaultOpen: false
};

const withMenuState = MenuComponentWithoutState => {
  const MenuComponent = ({ showDelay, hideDelay, defaultOpen, ...props }) => (
    <MenuState showDelay={showDelay} hideDelay={hideDelay} defaultOpen={defaultOpen}>
      {menuState => <MenuComponentWithoutState menuState={menuState} {...props} />}
    </MenuState>
  );
  MenuComponent.propTypes = {
    showDelay: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    hideDelay: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    defaultOpen: PropTypes.bool
  };

  MenuComponent.defaultProps = {
    showDelay: 0,
    hideDelay: 0,
    defaultOpen: false
  };

  return MenuComponent;
};

export default withMenuState;
