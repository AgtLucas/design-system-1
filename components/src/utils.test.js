import {css} from 'styled-components';
import { lineHeightRatio, fontMetrics } from './utils';

const fontSizeMd = 16;
const fontSizeSm = 10;
const lineHeightMd = 24;
const lineHeightSm = 20;
const _ = undefined;

const theme = {
  font: {
    size: { 
      medium: fontSizeMd,
      small: fontSizeSm,
    },
    lineHeight: { 
      target: { 
        medium: lineHeightMd,
        small: lineHeightSm,
      }
    }
  }
}
describe('utils', () => {
  describe('lineHeightRatio', () => {
    describe('given a font size and line height', () => {
      it('calculates a line-height ratio', () => {
        expect(lineHeightRatio('small', 'small')({ theme })).toEqual(lineHeightSm/fontSizeSm);
      });
    });
    describe('given a font size and no line height', () => {
      it('calculates a line-height ratio with a medium line height', () => {
        expect(lineHeightRatio('small')({ theme })).toEqual(lineHeightMd/fontSizeSm);
      });
    });
    describe('given a line height and no font size', () => {
      it('calculates a line-height ratio with a medium font size', () => {
        expect(lineHeightRatio(_, 'small')({ theme })).toEqual(lineHeightSm/fontSizeMd);
      });
    });
    describe('when only the inner function is called', () => {
      it('returns a function', () => {
        expect(() => lineHeightRatio(_, 'small')).toBeInstanceOf(Function);
      });
    });
    describe('when no theme is provided', () => {
      it('returns a function', () => {
        expect(() => lineHeightRatio(_, 'small')()).toThrow("Cannot read property 'theme' of undefined");
        expect(() => lineHeightRatio(_, 'small')({})).toThrow("Cannot read property 'font' of undefined");
      });
    });
  });

  describe('fontMetrics', () => {
    describe('given a font size and line height', () => {
      it('matches snapshot', () => {
        expect(fontMetrics('small', 'small')({ theme })).toMatchSnapshot();
      });
    });
    describe('given a font size and no line height', () => {
      it('matches snapshot', () => {
        expect(fontMetrics('small')({ theme })).toMatchSnapshot();
      });
    });
    describe('given a line height and no font size', () => {
      it('matches snapshot', () => {
        expect(fontMetrics(_, 'small')({ theme })).toMatchSnapshot();
      });
    });
    describe('when only the inner function is called', () => {
      it('returns a function', () => {
        expect(() => fontMetrics(_, 'small')).toBeInstanceOf(Function);
      });
    });
    describe('when no theme is provided', () => {
      it('returns a function', () => {
        expect(() => fontMetrics(_, 'small')()).toThrow("Cannot read property 'theme' of undefined");
        expect(() => fontMetrics(_, 'small')({})).toThrow("Cannot read property 'font' of undefined");
      });
    });
  });
});