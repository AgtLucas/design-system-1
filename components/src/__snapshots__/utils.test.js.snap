// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils fontMetrics given a font size and line height matches snapshot 1`] = `
Array [
  "
  font-size: ",
  "10",
  "px;
  line-height: ",
  [Function],
  ";
",
]
`;

exports[`utils fontMetrics given a font size and no line height matches snapshot 1`] = `
Array [
  "
  font-size: ",
  "10",
  "px;
  line-height: ",
  [Function],
  ";
",
]
`;

exports[`utils fontMetrics given a line height and no font size matches snapshot 1`] = `
Array [
  "
  font-size: ",
  "16",
  "px;
  line-height: ",
  [Function],
  ";
",
]
`;
